File: 21-Find if Path Exists in Graph/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        unordered_map<int, vector<int>> graph;
        for (auto& edge : edges) {
            int src = edge[0];
            int dst = edge[1];
            graph[src].push_back(dst);
            graph[dst].push_back(src);
        }

        queue<int> q;
        unordered_set<int> vis;

        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            if (curr == destination) return true;
            for (int neighbor : graph[curr]) {
                if (vis.find(neighbor) == vis.end()) {
                    vis.insert(neighbor);
                    q.push(neighbor);
                }
            }
        }

        return false;
    }
};

File: 21-Find if Path Exists in Graph/solution.py
import collections
from collections import deque

class Solution:
    def validPath(self, n, edges, source, destination):
        graph = collections.defaultdict(list)
        for edge in edges:
            src, dst = edge[0], edge[1]
            graph[src].append(dst)
            graph[dst].append(src)

        q = deque()
        vis = set()
        q.append(source)
        vis.add(source)

        while q:
            curr = q.popleft()
            if curr == destination:
                return True
            for neighbor in graph[curr]:
                if neighbor not in vis:
                    q.append(neighbor)
                    vis.add(neighbor)
        return False